torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
+    bw_compiler, get_bwd = _make_get_graph_compiler()
+    fn_compiled = aot_function(
+        fn,
+        fw_compiler=fw_compiler,
+        bw_compiler=bw_compiler,
+        partition_fn=partition_fn)
+    _, bwd_out_spec = tree_flatten((fake_carry_pytree, fake_x_pytree))
+    out = fn_compiled(fake_carry_pytree, fake_x_pytree)
+    out, fwd_out_spec = tree_flatten(out)
+    torch.autograd.backward(out, tree_map(lambda v: torch.ones_like(v), out))
+    # How many outputs out of the fwd_graph is actually outputs of `fn`, and not
+    # intermediate activations.
+    num_out = len(out)
+
+  fwd_graph = get_fwd()
+  bwd_graph = get_bwd()
+
+  def forward(carry, x):
+    flat_carry, _ = tree_flatten(carry)
+    flat_x, _ = tree_flatten(x)
+    out = fwd_graph(*flat_carry, *flat_x)
+    actual_out, activations = split(out, num_out)
+    carry, y = tree_unflatten(actual_out, fwd_out_spec)
+    y = (y, activations)
+    return carry, y
+
+  def backward(carry, x):
+    grad_new_carry, _ = tree_flatten(carry)
+    (grad_y, activations) = x
+    grad_y, _ = tree_flatten(grad_y)
+    out = bwd_graph(*activations, *grad_new_carry, *grad_y)
+    grad_carry, grad_x = tree_unflatten(out, bwd_out_spec)
+    return grad_carry, grad_x
+
+  return forward, backward
+
+
+def _make_get_graph_compiler():
+  """
+  Creates a compiler that records the graph, and a getter
+  function to retrieve them.
+  """
+  graph: List[Optional[torch.fx.GraphModule]] = [None]
+
+  def forward_comp(fx_module: torch.fx.GraphModule, _):
+    assert graph[0] is None
+    graph[0] = fx_module
+    return make_boxed_func(fx_module)
+
+  def get_graph():
+    g = graph[0]
+    assert g is not None
+    return g
+
+  return forward_comp, get_graph
+
+
+@pytreeify
+class Scan(torch.autograd.Function):
+
+  @staticmethod
+  def forward(ctx, forward, backward, init, xs):
+    # Forward pass, save activations for backward
+    ctx._backward = backward
+    with torch.no_grad():
+      carry, ys = _scan_impl_pytree(forward, init, xs)
+      ys, activations = ys
+    ctx.save_for_backward(*activations)
+    return carry, ys
+
+  @staticmethod
+  def backward(ctx, grad_carry, grad_ys):
+    activations = ctx.saved_tensors
+    backward = ctx._backward
+    with torch.no_grad():
+      # Reverse loop to propagate gradients from last iteration to first.
+      grad_init, grad_xs = _scan_impl_pytree(
+          backward, grad_carry, (grad_ys, activations), reverse=True)
+
+    # 2 comes from `pytreeify`. Other 2 are the `forward`, `backward` arguments
+    # passed to the `forward` method.
+    gradient_skipper = _GradientSkipper(ctx, starting_index=4)
+    grad_init = gradient_skipper.process(grad_init)
+    grad_xs = gradient_skipper.process(grad_xs)
+    return None, None, grad_init, grad_xs
+
+
+class _GradientSkipper:
+  """
+  Helper object to replace zero gradients with `None` gradients, if
+  the `needs_input_grad` value for the corresponding input tensor is False.
```

Comment:
why is this needed?

Reply:
It's not! :)

I think it was needed at one point where I didn't handle `None` value in PyTrees nicely. Now it's not needed.

Reply:
Removed